File: JavaScript/Sorting-Algorithms/InsertionSort.js
/*
******************* Insertion Sort Algorithm *********
Builds up the sort by gradually creation a larger left half which is always sorted
# Example:
[5, 4, 3, 2, 1]
[3, 4, 4, 1, 2]
[3, 4, 5, 1, 2]
[1, 3, 4, 5, 2]
[1, 2, 3, 4, 5]
*/

/*
Pseudo Code
1. Start by picking the second element in the array
2. Now compare the second element with the one before it and swap if necessary
3. Continue to the next element and if it is in the incorrect order, iterate through the sorted portion(i.e. the left side) to place the element in the correct place
4. Repeat until the array is sorted
*/

function insertionSort(arr) {
  for (let i = 1; i < arr.length; i++) {
    let currentVal = arr[i];
    for (let j = i - 1; j >= 0 && arr[j] > currentVal; j--) {
      arr[j + 1] = arr[j];
    }
    arr[j + 1] = currentVal;
  }
  return arr;
}

insertionSort([2, 1, 9, 76, 4]);

File: PYTHON/algorithms/insertionSort.py
# Python program for insertion sort
  
#function for the insertion sort
def insertionsort(A):
    #getting the length of the array
    n=len(A) 
    #traverse through 1 to length of the array
    for j in range(1,n):
        key=A[j]
        i=j-1
        #compare key with each element on the left of it
        #until an element smaller than it is found
        while(i>=0 and A[i]>key):
            A[i+1]=A[i]
            i=i-1
        #place key after the element that is smaller than it.
        A[i+1]=key
        
# Main program
print("array before insertion sort")
A = [22, 12, 43, 2, 60]
print(A)
insertionsort(A)

print("array after insertion sort")
insertionsort(A)
print(A)
